src/app/PendingResponseTracker.h

```diff
@@ -0,0 +1,89 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <lib/core/CHIPError.h>
+#include <lib/core/Optional.h>
+
+namespace chip {
+namespace app {
+
+/**
+ * @brief Interface for tracking responses to outbound InvokeRequests.
+ *
+ * This interface enables clients to:
+ *   * Verify that received responses correspond to issued InvokeRequests.
+ *   * Detect outstanding responses after the server indicates completion, helpful for identifying response omissions.
+ */
+class PendingResponseTracker
+{
+public:
+    virtual ~PendingResponseTracker() = default;
+
+    /**
+     * @brief Adds a pending response to the tracker. Tracked using CommandReference.
+     *
+     * @param [in] aCommandRef The CommandReference associated with the response for which we are
+     *             waiting.
+     * @return CHIP_NO_ERROR on success.
+     * @return CHIP_ERROR_INVALID_ARGUMENT if aCommandRef is already being tracked.
+     */
+    virtual CHIP_ERROR AddPendingResponse(uint16_t aCommandRef) = 0;
+
+    /**
+     * @brief Removes the pending response from the tracker as a response is received.
+     *
+     * Responses are tracked using CommandReference.
+     *
+     * @param [in] aCommandRef The CommandReference of the response that is no longer pending.
+     * @return CHIP_NO_ERROR on success.
+     * @return CHIP_ERROR_KEY_NOT_FOUND if aCommandRef is not found in the tracker. This indicates the
+     *             tracker was not waiting for a response associated with this CommandReference.
+     */
+    virtual CHIP_ERROR ResponseReceived(uint16_t aCommandRef) = 0;
+
+    /**
+     * @brief Indicates whether a response associated with a CommandReference is currently pending.
+     *
+     * @param [in] aCommandRef The CommandReference associated with the response to check if response
+     *             is pending.
+     * @return True if the response is pending, False otherwise.
```

Comment:
nit: some of these doc-comments could be made shorter - this return is obvious for example, so are all the `returns CHIP_NO_ERROR on success`, and overall `Response` and `PendingResponse` is very repeated (all except count):

```cpp
/// start tracking the given `commandRef`
///
/// returns CHIP_ERROR_INVALID_ARGUMENT if already tracked
virtual CHIP_ERROR Add(uint16_t commandRef) = 0;

/// Removes the tracking for the given command reference
///
/// returns CHIP_ERROR_KEY_NOT_FOUND if this value is not currently tracked
virtual CHIP_ERROR Remove(uint16_t commandRef) = 0;

/// check if the given `commandRef` is being tracked
virtual bool IsTracked(uint16_t commandRef);
```

.... overall this seems to behave a lot like a Set with explicity CHIP_ERROR semantics. Any reason not to use methods named as a set (Add/Remove/Empty/Size)?